### PR TITLE
Update for modern standards

### PR DESCRIPTION
**Pull request by CAD97**

### CODE REVIEW

**Comment by CAD97** [comment]
Hey @CeleritasCelery, would you mind spot-checking these documentation improvements before I publish the update? I've learned a lot since 2019 and want this documentation to reflect the current status of things as well as possible.

File: README.md [hunk 1]
```diff
@@ -32,30 +28,52 @@ hygiene hidden `impl Drop` can be used by any API that would normally use a clos
 argument, such as crossbeam's scoped threads, to make the containing scope the safe,
 wrapped scope if they so desire the trade-off of macro versus closure indentation.
 
+It's important to note that lifetimes are only trusted to cary lifetime brands when they
```
Comment (CeleritasCelery) [inline README.md]: cary -> carry

File: README.md [hunk 2]
```diff
@@ -32,30 +28,52 @@ hygiene hidden `impl Drop` can be used by any API that would normally use a clos
...
+are in fact invariant. Variant lifetimes, such as `&'a T`, can be shrunk to fit the expected
```
Comment (CeleritasCelery) [inline README.md:32]: "Variant lifetimes" -> "Covariant lifetimes"

"Variant" itself is not a category of lifetimes.

It might be good to add a link to the [docs](https://doc.rust-lang.org/nomicon/subtyping.html) as well

File: README.md [hunk 3]
```diff
@@ -105,16 +125,23 @@ fn main() {
 }
 ```
 
-Larger lifetimes can still shrink to unify with `'id`. What all this means is that you
-can't just trust any old `'id` lifetime floating around. It has to be carried in a
-trusted carrier; one that can't be created from an untrusted lifetime carrier.
+Larger variant lifetimes can still shrink to unify with `'id`. What all this
```
Comment (CeleritasCelery) [inline README.md]: variant -> covariant

File: src/lib.rs [hunk 4]
```diff
@@ -1,20 +1,37 @@
...
-//! [`Id`], which can be stored in structures to uniquely "brand" them. A different
-//! invocation of the macro will produce a new lifetime that cannot be unified.
-//! These types have no safe way to construct them other than via
-//! [`make_guard!`].
+//! unique among other trusted carriers. When you call [`make_guard!`] to make a
+//! unique lifetime, the macro creates a [`Guard`] to hold it. This guard can be
+//! converted `into` an [`Id`], which can be stored in structures to uniquely
+//! "brand" them. A different invocation of the macro will produce a new
+//! lifetime that cannot be unified. These types have no safe way to construct
+//! them other than via [`make_guard!`] or `unsafe` code.
```
Comment (CeleritasCelery) [inline src/lib.rs]: nitpick: unsafe code is not a "safe way" to construct something. So maybe we just say:

```
The only way to construct these types is with [`make_guard!`] or `unsafe` code.
```

File: src/lib.rs [hunk 5]
```diff
@@ -1,20 +1,37 @@
...
 //! Struct(a.into());
 //! ```
+//!
+//! This is the concept of "generative" lifetime brands. `Guard` and `Id` are
+//! [invariant](https://doc.rust-lang.org/nomicon/subtyping.html#variance) over
+//! their lifetime parameter, meaning that it is never valid to substitute or
+//! otherwise coerce `Id<'a>` into `Id<'b>`, for *any* `'a` or `'b`, *including*
+//! the `'static` lifetime.
+//!
+//! Any invariant lifetime can be "trusted" to cary a brand, but when using this
```
Comment (CeleritasCelery) [inline src/lib.rs]: > Any invariant lifetime can be "trusted" to cary a brand

I don't think this is necessarily true. It's possible to create invariant lifetimes that unify. Which means they **can't** be trusted to carry a brand. That property is true of lifetimes returned by `make_guard`, but not true in general.

cary -> carry

> `Guard` and `Id` are invariant over their lifetime parameter, meaning that it is never valid to substitute or otherwise coerce `Id<'a>` into `Id<'b>`

related to the above point, it is not the invariance *itself* that gives them this property, but rather the way `generativity` implements the scoped drop guards that prevents unification. Invariance is a necessary but not sufficient condition for "generative lifetimes" to work.

File: src/lib.rs [hunk 6]
```diff
@@ -26,8 +43,16 @@ pub struct Id<'id> {
 }
 
 impl<'id> Id<'id> {
-    // Do not use this function; use the `make_guard!` macro instead.
-    #[doc(hidden)]
+    /// Construct an `Id` with an unbound lifetime.
```
Comment (CeleritasCelery) [inline src/lib.rs]: unbound -> unbounded

See this [doc](https://doc.rust-lang.org/nomicon/unbounded-lifetimes.html)

File: src/lib.rs [hunk 7]
```diff
@@ -26,8 +43,16 @@ pub struct Id<'id> {
...
+    ///
+    /// You should not need to use this function; use [`make_guard!`] instead.
+    ///
+    /// # Safety
+    ///
+    /// This creates an unbound invariant lifetime that people are allowed to
+    /// assume means it was derived from their generative brand. This is the
+    /// "I know what I'm doing" button; restrict the lifetime to a known brand
+    /// immediately to avoid accidentally introducing unsoundness potential.
```
Comment (CeleritasCelery) [inline src/lib.rs]: To make the safe usage requirements clearer, I would use wording similar to this:

> `Id` holds an invariant lifetime that must be derived from a generative brand. Using this function directly is the "I know what I'm doing" button. Restrict the lifetime to a known brand immediately to avoid introducing unsoundness.

File: src/lib.rs [hunk 8]
```diff
@@ -60,10 +84,18 @@ pub struct Guard<'id> {
...
+    /// Construct a `Guard` with an unbound lifetime.
+    ///
+    /// You should not need to use this function; use [`make_guard!`] instead.
+    ///
+    /// # Safety
+    ///
+    /// This creates an unbound invariant lifetime that people are allowed to
+    /// assume means it was derived from their generative brand. This is the
+    /// "I know what I'm doing" button; restrict the lifetime to a known brand
+    /// immediately to avoid accidentally introducing unsoundness potential.
```
Comment (CeleritasCelery) [inline src/lib.rs]: Same comments for `Guard` as for `Id`.

File: README.md [hunk 9]
```diff
@@ -105,16 +125,23 @@ fn main() {
...
+means is that you can't just trust any old `'id` lifetime floating around. It
+has to be carried in a trusted carrier, and one that can't be created from an
+untrusted lifetime carrier.
```
Comment (CeleritasCelery) [inline README.md]: > and one that can't be created from an untrusted lifetime carrier.

I would just remove this extra condition because part of the contract of a "trusted carrier" is that it *was not* created from an untrusted lifetime, otherwise you have UB. When I see a trusted carrier it is safe to assume to was constructed properly.

 But if we want to keep this clause I would rephrase it as: 

> and one that wasn't created from an untrusted lifetime.

Because technically you *can* always create a carrier from an untrusted lifetime with unsafe code.

File: README.md [hunk 10]
```diff
@@ -32,30 +28,52 @@ hygiene hidden `impl Drop` can be used by any API that would normally use a clos
...
-- `#[inline(always)]`: This is a micro-optimization not required for safety. This makes it easier
-  for the optimizer to optimize out the `_guard`'s drop implementation.
+- New unique type per macro invocation: this is merely to avoid having a type in
+  the public API, and such that the compiler emits slightly more useful error
+  messages for lifetime errors.
+- `#[inline(always)]`: This is a micro-optimization not required for safety.
+  This makes it easier for the optimizer to optimize out the `_guard`'s drop
+  implementation.
+- `make_guard` is created from `&'id Id<'id>` but only holds `Id<'id>`. While
+  the reference is required to uniquify the lifetime (see below), only `Id<'id>`
```
Comment (CeleritasCelery) [inline README.md:77]: where should I be looking for the "see below" reference?

File: README.md [hunk 11]
```diff
@@ -32,30 +28,52 @@ hygiene hidden `impl Drop` can be used by any API that would normally use a clos
...
-See also #1 for further discussion.
+### Disclaimer
+
+This relies on dead code (the empty drop) to impact borrow checking.
+Theoretically, a smarter CFG based borrow checker (i.e. NLL/polonius) *could*
+utilize the fact that this is dead code to remove this restriction, but this is
+*very* unlikely; the current (as of Rust 2021) NLL borrow checker requires dead
+code to be lifetime-correct, and this code isn't *dead* dead, as in, it *runs*,
+it just doesn't actually do anything other than impact lifetime solving. If you
+want to discuss the proof of correctness, the place to do so is issue #1.
```
Comment (CeleritasCelery) [inline README.md]: I would make `#1` an actual link to the issue.

**Review by CeleritasCelery — COMMENTED** [review]
I like the updated docs! Feel free to use or reject my suggestions as you see fit.

File: README.md (hunk 2 above)
Comment (CAD97) [reply README.md:32]: The variance docs are linked above. It is correct that only covariant lifetimes can be shrunk, but a contravariant lifetime could in theory grow to fit the expected lifetime as well. I'm using "variant" here as "not invariant".

File: README.md (hunk 11 above)
Comment (CAD97) [reply README.md]: GitHub automatically linkifies #issue-number, but I suppose that yeah, it's nice to link it for e.g. crates.io.

File: README.md (hunk 10 above)
Comment (CAD97) [reply README.md:77]: The ``### `'id` is unique`` heading.

**Review by CAD97 — COMMENTED** [review]
Thanks for the notes @CeleritasCelery!
